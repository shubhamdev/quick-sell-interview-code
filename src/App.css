:root {
  --primary-color: #b24242;
  --secondary-color: #fbe8e8;
  --tertiary-color: #ffffff;
  --background-color: #f9f9f9;
  --shadow-color: #cccccc;
}

* {
  font-size: 21px;
  font-weight: 900;
  color: var(--primary-color);
}

.App {
  background-color: var(--background-color);
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
}

.wrapper {
  padding: 0 40px;
  background-color: var(--tertiary-color);
  box-shadow: 0 0 8px -2px var(--shadow-color);
}

.loader {
  font-size: 10px;
  margin: 0 10px 0 0;
  text-indent: -9999em;
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background: #ffffff;
  position: relative;
  -webkit-animation: load3 1.4s infinite linear;
  animation: load3 1.4s infinite linear;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}
.loader:before {
  width: 50%;
  height: 50%;
  background: #000000;
  border-radius: 100% 0 0 0;
  position: absolute;
  top: 0;
  left: 0;
  content: "";
}
.loader:after {
  background: #ffffff;
  width: 75%;
  height: 75%;
  border-radius: 50%;
  content: "";
  margin: auto;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}
@-webkit-keyframes load3 {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
@keyframes load3 {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.loading-state {
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  align-items: center;
  height: 40px;
}

.wrapper p {
  color: #000000;
  font-weight: normal;
  font-size: 12px;
}
