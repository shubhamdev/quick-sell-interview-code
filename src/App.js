import React, { useState, useEffect, useRef } from "react";
import { debounce } from "lodash-es";

import Counter from "./components/Counter";
import useCounter from "./hooks/useCounter";

import "./App.css";

const BASE_URL = "https://interview-8e4c5-default-rtdb.firebaseio.com";

const PUT_URL = BASE_URL + "/front-end.json";
const GET_URL = BASE_URL + "/front-end/quickSellCounter.json";

const apiCall = async (url = "", method = "GET", body) => {
  const options = {
    method,
    headers: {
      "Content-Type": "application/json",
    },
    ...(body ? { body: JSON.stringify(body) } : {}),
  };

  try {
    const res = await fetch(url, options).then(res => res.json());
    return res;
  } catch (e) {
    throw e;
  }
};

function App({ initialValue = 1, max = 1000 }) {
  const [counter, dispatchCounter] = useCounter({ count: initialValue, max });
  const [isSaving, setIsSaving] = useState(false);
  const counterValueFromGet = useRef(initialValue);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const value = await apiCall(GET_URL);
        if (value && value !== null) {
          counterValueFromGet.current = value;
          dispatchCounter({
            type: "absolute",
            value,
          });
        }
      } catch (e) {
        console.log(e);
      }
    };
    fetchData();
  }, [dispatchCounter]);

  const handleIncrement = () => dispatchCounter({ type: "increment" });
  const handleDecrement = () => dispatchCounter({ type: "decrement" });
  const onChange = e =>
    dispatchCounter({ type: "absolute", value: e.target.value });

  const insertCounterDebounced = useRef(
    debounce(async count => {
      try {
        setIsSaving(true);
        await apiCall(PUT_URL, "PUT", { quickSellCounter: count });
      } catch (e) {
        console.log("Unable to save the Counter Value", e);
      } finally {
        setIsSaving(false);
      }
    }, 500)
  ).current;

  useEffect(() => {
    if (counterValueFromGet.current !== counter.count)
      insertCounterDebounced(counter.count);
  }, [counter.count, insertCounterDebounced]);

  return (
    <div className="App">
      <div className="wrapper">
        <div
          className="loading-state"
          style={isSaving ? {} : { visibility: "hidden" }}
        >
          <div className="loader"></div>
          <p>Saving counter value</p>
        </div>
        <Counter
          count={counter.count}
          max={counter.max}
          handleDecrement={handleDecrement}
          handleIncrement={handleIncrement}
          onChange={onChange}
        />
        <p>Counter value : {counter.count}</p>
      </div>
    </div>
  );
}

export default App;
