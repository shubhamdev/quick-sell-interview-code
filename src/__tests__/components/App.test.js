import { render } from "@testing-library/react";
import pretty from "pretty";

import App from "../../App";

test("renders app", () => {
  const { container } = render(<App />);

  expect(container).toBeInTheDocument();

  expect(pretty(container.innerHTML)).toMatchSnapshot();
});
