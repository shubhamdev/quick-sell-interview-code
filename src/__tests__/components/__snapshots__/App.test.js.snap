// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders app 1`] = `
"<div class=\\"App\\">
  <div class=\\"wrapper\\">
    <div class=\\"counter\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 1px 0px 0px; border-style: solid;\\">-</button><input type=\\"number\\" class=\\"count-input\\" aria-label=\\"Count\\" max=\\"1000\\" value=\\"1\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 0px 0px 1px; border-style: solid;\\">+</button></div>
  </div>
</div>"
`;
